daemon/anonymous-statistics.sh

```diff
@@ -0,0 +1,157 @@
+#      ACTION_RESULT  -- OK or FAILED
+# - FATAL       - netdata exited due to a fatal condition
+#      ACTION_RESULT  -- error code
+#      ACTION_DATA    -- error message
+# - START       - netdata started
+# - INSTALL     - installation action
+#      ACTION_DATA    -- installation method from /var/lib/netdata/.installation
+#      ACTION_RESULT  -- OK or FAILED
+
+ACTION="${1}"
+ACTION_RESULT="${2}"
+ACTION_DATA="${3}"
+
+# -------------------------------------------------------------------------------------------------
+# check opt-out
+
+OPTOUT=
+if [ -z "${NETDATA_CONFIG_DIR}" ]; then
+    for dir in "$(netdata -W get global 'config directory' '/etc/netdata' 2>/dev/null)" "/etc/netdata" "/opt/netdata/etc/netdata"; do
+        if [ -f "${dir}/.opt-out-from-anonymous-statistics" ]; then
+            OPTOUT=all
+            break
+        fi
+    done
+else
+    if [ -f "${NETDATA_CONFIG_DIR}/.opt-out-from-anonymous-statistics" ]; then
+        OPTOUT=all
+    fi
+fi
+
+if [ "${OPTOUT}" = "all" ] ; then
+    exit 0
+fi
+
+# -------------------------------------------------------------------------------------------------
+# detect the operating system
+
+OS_DETECTION=
+NAME=
+VERSION=
+VERSION_ID=
+ID=
+ID_LIKE=
+
+if [ -f "/etc/os-release" ]; then
+    OS_DETECTION="/etc/os-release"
+    eval "$(grep -E "^(NAME|ID|ID_LIKE|VERSION|VERSION_ID)=" </etc/os-release)"
+elif [ -f "/etc/lsb-release" ]; then
+    OS_DETECTION="/etc/lsb-release"
+    DISTRIB_ID=
+    DISTRIB_RELEASE=
+    DISTRIB_CODENAME=
+    eval "$(grep -E "^(DISTRIB_ID|DISTRIB_RELEASE|DISTRIB_CODENAME)=" </etc/lsb-release)"
+    NAME="${DISTRIB_ID}"
+    VERSION="${DISTRIB_RELEASE}"
+    ID="${DISTRIB_CODENAME}"
+elif [ ! -z "$(command -v lsb_release 2>/dev/null)" ]; then
+    OS_DETECTION="lsb_release"
+    NAME="$(lsb_release -i 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
+    VERSION_ID="$(lsb_release -f 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
+    VERSION="$(lsb_release -c 2>/dev/null | sed "s/[[:space:]]*:[[:space:]]*/:/g" | cut -d ':' -f 2)"
+fi
+
+# -------------------------------------------------------------------------------------------------
+# detect the kernel
+
+KERNEL_NAME="$(uname -s)"
+KERNEL_VERSION="$(uname -r)"
+ARCHITECTURE="$(uname -m)"
+
+# -------------------------------------------------------------------------------------------------
+# detect the virtualization
+
+VIRTUALIZATION="unknown"
+VIRT_DETECTION="none"
+
+if [ ! -z "$(command -v systemd-detect-virt 2>/dev/null)" ]
+then
+    VIRTUALIZATION="$(systemd-detect-virt)"
+    VIRT_DETECTION="systemd-detect-virt"
+else
+    # /proc/1/sched exposes the host's pid of our init !
+    # http://stackoverflow.com/a/37016302
+    pid=$( head -n 1 </proc/1/sched 2>/dev/null | { IFS='(),#:' read -r _unused pid _unused _unused; echo $pid; } )
+    if [ ! -z "${pid}" ]; then
+        pid=$(( pid + 0 ))
+        [ ${pid} -ne 1 ] &&
```

Comment:
To be fair this:
```
pid=$( head -n 1 </proc/1/sched 2>/dev/null | { IFS='(),#:' read -r _unused pid _unused _unused; echo $pid; } )
echo $pid
```

Will return `1` when running inside a container or outside of a container, because we are asking for PID 1: "/proc/**1**/sched"

Reply:
I think I got what was the intention and if I am correct, then this whole section should be written as:
```sh
process=$( head -n 1 </proc/1/sched 2>/dev/null | { IFS='(),#:' read -r process _unused _unused _unused; echo $process; } )

if [ "${process}" == "netdata" ]; then
   VIRTUALIZATION="container"
   VIRT_DETECTION="pid"
fi
```

Reply:
Or even simpler:
```sh
IFS='(,' read -r process _unused </proc/1/sched
if [ "${process}" == "netdata" ]; then
   VIRTUALIZATION="container"
   VIRT_DETECTION="pid"
fi
```

Reply:
I included this one as well.